src/ray/core_worker/store_provider/memory_store/memory_store.h

```diff
@@ -39,6 +39,15 @@ class CoreWorkerMemoryStore {
   Status Get(const std::vector<ObjectID> &object_ids, int num_objects, int64_t timeout_ms,
              bool remove_after_get, std::vector<std::shared_ptr<RayObject>> *results);
 
+  /// Asynchronously get an object from the object store. The object will not be removed
+  /// from storage after GetAsync (TODO(ekl): integrate this with object GC).
+  ///
+  /// \param[in] object_id The object id to get.
+  /// \param[in] callback The callback to run with the reference to the retrieved
+  ///            object value once available.
+  void GetAsync(const ObjectID &object_id,
```

Comment:
I'm a bit concerned about the approach of just attaching callbacks to an object ID, without any information about what task the callback is going to launch. It makes it hard to do stuff like cancel one particular task, since we'd have to go find all the callbacks that it attached. No need to fix it right now, but any thoughts on how to make this better?

Reply:
It also occurs to me that it'll be difficult to tell how many tasks are queued in the worker due to missing data dependencies. I guess we'll need another hashmap at some point to fix it.

Reply:
I see, we will definitely need to expose this as a metric at some point. I added a TODO in LocalDependencyResolver to expose metrics.

For cleanup -- I'm not too concerned, since the amount of "junk" is bounded and will eventually clean itself up even if we cancel tasks, since gRPC guarantees error returns. We just need to make sure the errors are written into the store.

Reply:
Hmm I was thinking of cases where you want to cancel a task that hasn't been sent yet, but I guess we don't even have that functionality yet in master. Okay, seems fine as long as we have a metric later to measure number of pending tasks.